### DISCUSSION TITLE
SB Reports - metrics issue

### DISCUSSION THREAD

**Discussion by michalaleksandrowicz** [denisneuf/python-amazon-ad-api#128] [General]
Dear Friends,

As a home made hobbyist developer, I have been trying to pull SB reports to my database using python. It worked magic for SP and SD, now I have struggled for the past week with SB and the issue was metrics. Whenever I tried to pull in more than one metric, I was getting for example this one:

'payload': {'code': '422',
             'details': 'Unrecognized metric:  campaignRuleBasedBudget\n'
                        ' at [Source:',
             'requestId': 'W0DEWDAXMT3DVB8751DZ'}}

I had metrics written like this in the data dictionary:

"metrics": "campaignName, currency, campaignId, campaignRuleBasedBudget"

And I could not get it work at all. Always getting Unrecognized Metric error.

Then finally Friday evening, today I try having them written without " ", after commas, like this:

"metrics": "campaignName,currency,campaignId,campaignRuleBasedBudget"

Sounds crazy, i know. But it is working now. Curious though if anyone knows why this is the case? For SD metrrics with " " after commas worked fine. Thanks!

**Comment by denisneuf** [comment]
Hi michalaleksandrowicz,
not surprised at all, they may have different monkey teams for sp, sd and sb api endpoints.
While sd clear you can see:
example: campaignName,campaignId,impressions,clicks,cost,attributedSales14d
A comma-separated list of the metrics to be included in the report.

Sponsored Display will ask you a string
metrics* | string

Sponsored Product V2 will be deprecate on [PLANNED DEPRECATION 3/30/2023]
metrics | ReportMetricsstring
A comma-separated list of the metrics to be included in the report. Each report type supports different metrics.

Sponsored Product V3 is moved to a new endpoint /reporting/reports and | **adProduct** (string): [required] Enum: The advertising product such as SPONSORED_PRODUCTS or SPONSORED_BRANDS.
And metrics disappear 

Sponsored Brands looks more like SPV2
| **metrics** (string) [optional] A comma-separated list of the metrics to be included in the report.

So we will add some notice in help to avoid space after commas that by the way is the standard I used in general. Your details error:

'details': 'Unrecognized metric: campaignRuleBasedBudget\n'

Have an ugly \n at the end that may point you to the problem.

Anyway thank you to share with us,

Kindest Regards,

Daniel.

**Reply by michalaleksandrowicz** [reply]
Yeah, thanks! The "\n" rang a bell once I solved the issue. If I had more experience, I would have discovered it earlier : /
